File: platform/kui/cli/Command.kt
package kui.cli

/**
 * Enumeration of all recognized KUI CLI commands as defined in Section 9 of the roadmap.
 */
enum class Command(
    val commandName: String,
    val description: String,
    val usage: String
) {
    NEW("new", "create project", "kui new <name>"),
    DOCTOR("doctor", "verify environment", "kui doctor"),
    INFO("info", "show project information", "kui info"),
    BUILD("build", "compile/package", "kui build [options]"),
    RUN("run", "build + install + launch", "kui run [options]"),
    INSTALL("install", "install current APK", "kui install"),
    LAUNCH("launch", "launch installed app", "kui launch"),
    TEST("test", "run tests", "kui test"),
    CLEAN("clean", "remove generated build output", "kui clean"),
    BENCH("bench", "benchmarks", "kui bench [benchmark-name]"),
    PROFILE("profile", "profiling report", "kui profile"),
    UI_TREE("ui-tree", "print runtime UI tree later", "kui ui-tree"),
    HELP("help", "show help for commands", "kui help [command]"),
    VERSION("version", "print version information", "kui version");

    companion object {
        fun fromString(name: String): Command? =
            values().firstOrNull { it.commandName.equals(name, ignoreCase = true) }

        /**
         * Computes closest command using Levenshtein distance for friendly diagnostics.
         */
        fun findClosest(name: String): Command? {
            return values()
                .map { it to levenshteinDistance(name.lowercase(), it.commandName) }
                .filter { it.second <= 2 }
                .minByOrNull { it.second }
                ?.first
        }

        private fun levenshteinDistance(s1: String, s2: String): Int {
            val dp = Array(s1.length + 1) { IntArray(s2.length + 1) }
            for (i in 0..s1.length) dp[i][0] = i
            for (j in 0..s2.length) dp[0][j] = j
            for (i in 1..s1.length) {
                for (j in 1..s2.length) {
                    val cost = if (s1[i - 1] == s2[j - 1]) 0 else 1
                    dp[i][j] = minOf(
                        dp[i - 1][j] + 1,
                        dp[i][j - 1] + 1,
                        dp[i - 1][j - 1] + cost
                    )
                }
            }
            return dp[s1.length][s2.length]
        }
    }
}

File: crates/kui-cli/src/cli/command.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Command {
    New,
    Doctor,
    Info,
    Build,
    Run,
    Install,
    Launch,
    Test,
    Clean,
    Bench,
    Profile,
    UiTree,
    Devices,
    Help,
    Version,
}

impl Command {
    pub fn all() -> &'static [Command] {
        &[
            Command::New,
            Command::Doctor,
            Command::Info,
            Command::Build,
            Command::Run,
            Command::Install,
            Command::Launch,
            Command::Test,
            Command::Clean,
            Command::Bench,
            Command::Profile,
            Command::UiTree,
            Command::Devices,
            Command::Help,
            Command::Version,
        ]
    }

    pub fn command_name(&self) -> &'static str {
        match self {
            Command::New => "new",
            Command::Doctor => "doctor",
            Command::Info => "info",
            Command::Build => "build",
            Command::Run => "run",
            Command::Install => "install",
            Command::Launch => "launch",
            Command::Test => "test",
            Command::Clean => "clean",
            Command::Bench => "bench",
            Command::Profile => "profile",
            Command::UiTree => "ui-tree",
            Command::Devices => "devices",
            Command::Help => "help",
            Command::Version => "version",
        }
    }

    pub fn description(&self) -> &'static str {
        match self {
            Command::New => "create project",
            Command::Doctor => "verify environment",
            Command::Info => "show project information",
            Command::Build => "compile/package",
            Command::Run => "build + install + launch",
            Command::Install => "install current APK",
            Command::Launch => "launch installed app",
            Command::Test => "run tests",
            Command::Clean => "remove generated build output",
            Command::Bench => "benchmarks",
            Command::Profile => "profiling report",
            Command::UiTree => "print runtime UI tree later",
            Command::Devices => "list connected devices",
            Command::Help => "show help for commands",
            Command::Version => "print version information",
        }
    }

    pub fn usage(&self) -> &'static str {
        match self {
            Command::New => "kui new <name>",
            Command::Doctor => "kui doctor",
            Command::Info => "kui info",
            Command::Build => "kui build [options]",
            Command::Run => "kui run [options]",
            Command::Install => "kui install",
            Command::Launch => "kui launch",
            Command::Test => "kui test",
            Command::Clean => "kui clean",
            Command::Bench => "kui bench [benchmark-name]",
            Command::Profile => "kui profile",
            Command::UiTree => "kui ui-tree",
            Command::Devices => "kui devices",
            Command::Help => "kui help [command]",
            Command::Version => "kui version",
        }
    }

    pub fn from_str(name: &str) -> Option<Command> {
        let lower = name.to_lowercase();
        Self::all().iter().copied().find(|c| c.command_name() == lower)
    }

    pub fn find_closest(name: &str) -> Option<Command> {
        let lower = name.to_lowercase();
        Self::all()
            .iter()
            .map(|&cmd| (cmd, levenshtein_distance(&lower, cmd.command_name())))
            .filter(|&(_, dist)| dist <= 2)
            .min_by_key(|&(_, dist)| dist)
            .map(|(cmd, _)| cmd)
    }
}

pub fn levenshtein_distance(s1: &str, s2: &str) -> usize {
    let s1_chars: Vec<char> = s1.chars().collect();
    let s2_chars: Vec<char> = s2.chars().collect();
    let m = s1_chars.len();
    let n = s2_chars.len();

    let mut dp = vec![vec![0usize; n + 1]; m + 1];

    for i in 0..=m {
        dp[i][0] = i;
    }
    for j in 0..=n {
        dp[0][j] = j;
    }

    for i in 1..=m {
        for j in 1..=n {
            let cost = if s1_chars[i - 1] == s2_chars[j - 1] { 0 } else { 1 };
            dp[i][j] = (dp[i - 1][j] + 1)
                .min(dp[i][j - 1] + 1)
                .min(dp[i - 1][j - 1] + cost);
        }
    }

    dp[m][n]
}
